File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Embeddings/EmbeddingsProto+Helpers.swift
//
//  EmbeddingsProto+Helpers.swift
//  RunAnywhere SDK
//
//  Ergonomic helpers for canonical Embeddings proto types.
//
//  defaults() / validate() factories live in
//  Generated/RAConvenience.swift, emitted by
//  idl/codegen/generate_swift_convenience.py from the rac_default /
//  rac_required / rac_min annotations in idl/embeddings_options.proto.
//

import Foundation

// MARK: - RAEmbeddingVector

extension RAEmbeddingVector {
    public func cosineSimilarity(with other: RAEmbeddingVector) -> Float {
        guard values.count == other.values.count, !values.isEmpty else { return 0 }
        var dot: Float = 0
        for i in 0..<values.count { dot += values[i] * other.values[i] }
        let aNorm = hasNorm ? norm : Self.l2(values)
        let bNorm = other.hasNorm ? other.norm : Self.l2(other.values)
        guard aNorm > 0 && bNorm > 0 else { return 0 }
        return dot / (aNorm * bNorm)
    }

    public func computeNorm() -> Float { Self.l2(values) }

    private static func l2(_ values: [Float]) -> Float {
        var sumSquares: Float = 0
        for value in values { sumSquares += value * value }
        return sumSquares.squareRoot()
    }
}

// MARK: - RAEmbeddingsResult

extension RAEmbeddingsResult {
    public var processingTime: TimeInterval { TimeInterval(processingTimeMs) / 1000.0 }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/Embeddings/EmbeddingsProto+Helpers.ts
/**
 * EmbeddingsProto+Helpers.ts
 *
 * Ergonomic helpers for canonical Embeddings proto types.
 *
 * Mirrors Swift `EmbeddingsProto+Helpers.swift`
 * (`RAEmbeddingVector.cosineSimilarity(with:)` / `computeNorm()`).
 */

import type { EmbeddingVector } from '@runanywhere/proto-ts/embeddings_options';

function l2(values: number[]): number {
  let sumSquares = 0;
  for (const value of values) sumSquares += value * value;
  return Math.sqrt(sumSquares);
}

/**
 * Cosine similarity between two embedding vectors.
 *
 * Mirrors Swift `RAEmbeddingVector.cosineSimilarity(with:)`
 * (EmbeddingsProto+Helpers.swift:18-26): returns 0 for mismatched lengths,
 * empty vectors, or zero norms; uses the backend-provided `norm` when
 * present and recomputes the L2 norm otherwise.
 */
export function cosineSimilarity(
  a: EmbeddingVector,
  b: EmbeddingVector
): number {
  if (a.values.length !== b.values.length || a.values.length === 0) return 0;
  let dot = 0;
  for (let i = 0; i < a.values.length; i++) {
    dot += a.values[i]! * b.values[i]!;
  }
  const aNorm = a.norm !== undefined ? a.norm : l2(a.values);
  const bNorm = b.norm !== undefined ? b.norm : l2(b.values);
  if (aNorm <= 0 || bNorm <= 0) return 0;
  return dot / (aNorm * bNorm);
}

/**
 * L2 norm of an embedding vector's values.
 * Mirrors Swift `RAEmbeddingVector.computeNorm()`.
 */
export function computeNorm(vector: EmbeddingVector): number {
  return l2(vector.values);
}
